irr/src/SkinnedMesh.cpp

```diff
@@ -299,92 +235,68 @@ void SkinnedMesh::prepareForSkinning()
-		for (u32 j = 0; j < Vertices_Moved[i].size(); ++j)
-			Vertices_Moved[i][j] = false;
-
-	// For skinning: cache weight values for speed
-	for (auto *joint : AllJoints) {
-		for (auto &weight : joint->Weights) {
-			const u16 buffer_id = weight.buffer_id;
-			const u32 vertex_id = weight.vertex_id;
-
-			weight.Moved = &Vertices_Moved[buffer_id][vertex_id];
-			weight.StaticPos = LocalBuffers[buffer_id]->getVertex(vertex_id)->Pos;
-			weight.StaticNormal = LocalBuffers[buffer_id]->getVertex(vertex_id)->Normal;
-		}
-	}
-
-	normalizeWeights();
-
 	for (auto *joint : AllJoints) {
 		joint->keys.cleanup();
 	}
 }
 
 void SkinnedMesh::calculateStaticBoundingBox()
 {
-	std::vector<std::vector<bool>> animated(getMeshBufferCount());
-	for (u32 mb = 0; mb < getMeshBufferCount(); mb++)
-		animated[mb] = std::vector<bool>(getMeshBuffer(mb)->getVertexCount());
-
-	for (auto *joint : AllJoints) {
-		for (auto &weight : joint->Weights) {
-			const u16 buffer_id = weight.buffer_id;
-			const u32 vertex_id = weight.vertex_id;
-			animated[buffer_id][vertex_id] = true;
-		}
-	}
-
 	bool first = true;
+	std::vector<bool> animated;
 	for (u16 mb = 0; mb < getMeshBufferCount(); mb++) {
-		for (u32 v = 0; v < getMeshBuffer(mb)->getVertexCount(); v++) {
-			if (!animated[mb][v]) {
-				auto pos = getMeshBuffer(mb)->getVertexBuffer()->getPosition(v);
-				if (!first) {
-					StaticPartsBox.addInternalPoint(pos);
-				} else {
-					StaticPartsBox.reset(pos);
-					first = false;
-				}
+		auto *buf = LocalBuffers[mb];
+		animated.clear();
+		animated.resize(buf->getVertexCount(), false);
+		if (buf->Weights) {
+			for (u32 vert_id : buf->Weights->animated_vertices.value()) {
+				animated[vert_id] = true;
+			}
+		}
+		for (u32 v = 0; v < buf->getVertexCount(); v++) {
+			if (animated[v])
+				continue;
+
+			auto pos = getMeshBuffer(mb)->getVertexBuffer()->getPosition(v);
+			if (!first) {
+				StaticPartsBox.addInternalPoint(pos);
+			} else {
+				StaticPartsBox.reset(pos);
+				first = false;
 			}
 		}
 	}
 }
 
 void SkinnedMesh::calculateJointBoundingBoxes()
 {
-	for (auto *joint : AllJoints) {
-		bool first = true;
-		for (auto &weight : joint->Weights) {
-			if (weight.strength < 1e-6)
-				continue;
-			auto pos = weight.StaticPos;
-			joint->GlobalInversedMatrix.value().transformVect(pos);
-			if (!first) {
-				joint->LocalBoundingBox.addInternalPoint(pos);
-			} else {
-				joint->LocalBoundingBox.reset(pos);
-				first = false;
+	std::vector<std::optional<core::aabbox3df>> joint_boxes(AllJoints.size());
+	for (auto *buf : LocalBuffers) {
+		const auto &weights = buf->Weights;
+		if (!weights)
+			continue;
+		for (u32 vert_id : weights->animated_vertices.value()) {
+			for (u16 j = 0; j < WeightBuffer::MAX_WEIGHTS_PER_VERTEX; j++) {
+				const auto joint_id = weights->getJointIds(vert_id)[j];
+				const auto *joint = AllJoints[joint_id];
+				const auto strength = weights->getWeights(vert_id)[j];
+				if (core::equals(strength, 0.0f))
```

Comment:
Style: `WeightBuffer::VertexWeights::skinVertex` is rather similar but with the following differences:
1. `strength` is called `weight` (I would prefer weight)
2. `auto` types are used here (I would prefer explicit types to ease static parsing (by eye))